### PR TITLE
Use classic synchronization so we can throw IOException

### PR DESCRIPTION
**Pull request by rnewson**

### CODE REVIEW

File: src/main/java/com/cloudant/search3/Search.java [hunk 1]
```diff
@@ -278,62 +280,58 @@ public void close() {
...
-                        commitIntervalSecs,
-                        TimeUnit.SECONDS);
-                LOGGER.info("Opened index {}", result);
+        synchronized(handlers) {
+            SearchHandler result = handlers.get(subspace);
+            if (result != null) {
                 return result;
-            } catch (final IOException e) {
-                LOGGER.catching(e);
-                return null;
```
Comment (tonysun83) [inline src/main/java/com/cloudant/search3/Search.java]: We no longer get an NPE. IIUC, you're letting the IOException go up the stack and where it is caught here:

https://github.com/cloudant-labs/search3-java/blob/9569ca2f746c152ec359574622480dda858eb443/src/main/java/com/cloudant/search3/Search.java#L302

Then in failedHandler you remove the index. Hence we see:

```
2019-11-20 22:01:19 WARN  Search:376 - Ignoring exception thrown while closing failed handler
java.io.FileNotFoundException: write.lock does not exist
	at com.cloudant.fdblucene.FDBDirectory.deleteFile(FDBDirectory.java:283) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.fdblucene.FDBLock.close(FDBLock.java:62) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.util.IOUtils.close(IOUtils.java:89) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.util.IOUtils.close(IOUtils.java:77) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.index.IndexWriter.rollbackInternalNoCommit(IndexWriter.java:2337) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.index.IndexWriter.rollbackInternal(IndexWriter.java:2268) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at org.apache.lucene.index.IndexWriter.rollback(IndexWriter.java:2261) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.FDBDirectorySearchHandler.close(FDBDirectorySearchHandler.java:82) ~[search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.failedHandler(Search.java:372) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.failedHandler(Search.java:362) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.execute(Search.java:303) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.Search.updateDocument(Search.java:227) [search3-1.0.0-SNAPSHOT.jar:?]
	at com.cloudant.search3.grpc.SearchGrpc$MethodHandlers.invoke(SearchGrpc.java:763) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:322) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:762) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) [search3-1.0.0-SNAPSHOT.jar:?]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) [search3-1.0.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
```

**Comment by tonysun83** [comment]
This was new that I haven't seen before from the previous runs:

```
Exception in thread "grpc-default-executor-549" java.lang.OutOfMemoryError: Java heap space
	at org.apache.lucene.util.packed.DirectMonotonicWriter.<init>(DirectMonotonicWriter.java:56)
	at org.apache.lucene.util.packed.DirectMonotonicWriter.getInstance(DirectMonotonicWriter.java:136)
	at org.apache.lucene.codecs.lucene80.Lucene80DocValuesConsumer.addTermsDict(Lucene80DocValuesConsumer.java:484)
	at org.apache.lucene.codecs.lucene80.Lucene80DocValuesConsumer.doAddSortedField(Lucene80DocValuesConsumer.java:472)
	at org.apache.lucene.codecs.lucene80.Lucene80DocValuesConsumer.addSortedField(Lucene80DocValuesConsumer.java:423)
	at org.apache.lucene.codecs.perfield.PerFieldDocValuesFormat$FieldsWriter.addSortedField(PerFieldDocValuesFormat.java:119)
	at org.apache.lucene.index.SortedDocValuesWriter.flush(SortedDocValuesWriter.java:163)
	at org.apache.lucene.index.DefaultIndexingChain.writeDocValues(DefaultIndexingChain.java:263)
	at org.apache.lucene.index.DefaultIndexingChain.flush(DefaultIndexingChain.java:138)
	at org.apache.lucene.index.DocumentsWriterPerThread.flush(DocumentsWriterPerThread.java:468)
	at org.apache.lucene.index.DocumentsWriter.doFlush(DocumentsWriter.java:555)
	at org.apache.lucene.index.DocumentsWriter.preUpdate(DocumentsWriter.java:402)
	at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:480)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1594)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1586)
	at com.cloudant.search3.FDBDirectorySearchHandler.updateDocument(FDBDirectorySearchHandler.java:204)
	at com.cloudant.search3.Search.lambda$updateDocument$5(Search.java:228)
	at com.cloudant.search3.Search$$Lambda$34/1700901186.accept(Unknown Source)
	at com.cloudant.search3.Search.execute(Search.java:301)
	at com.cloudant.search3.Search.updateDocument(Search.java:227)
	at com.cloudant.search3.grpc.SearchGrpc$MethodHandlers.invoke(SearchGrpc.java:763)
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:322)
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:762)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

File: src/main/java/com/cloudant/search3/Search.java (hunk 1 above)
Comment (rnewson) [reply src/main/java/com/cloudant/search3/Search.java]: That's right. This PR is simplifying how errors are handled and logged, it doesn't fix anything per se. Hopefully it makes it easier for us to see how to fix things, though.

**Comment by tonysun83** [comment]
+1

**Pull request merged**